Feature/1258 ad test core

bob-carpenter opened the pull request:
## Summary

This PR removes the old vectorized testing framework, updates all the tests that depend on the framework, and removes *a lot* of redundant tests in the `rev`, `fwd` and `mix` directories.  

The autodiff testing framework is documented in the code and [in an instructional wiki with examples](https://github.com/stan-dev/math/wiki/Automatic-Differentiation-Testing-Framework).

The new tests are more thorough and uncovered some inconsistencies between the vectorized versions and base primitive version in terms of error conditions, which are resolved in this PR.


## Tests

This is all new tests.  The 50 or so unary scalar functions that are vectorized using the vectorization framework are now automatically tested using the new autodiff test framework.  The tests are much simpler and also more thorough in testing error conditions and all derivatives up to third order.

## Side Effects

None.

## Checklist

- [x] Math issue #1258  

- [x] Copyright holder: Columbia University

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

seantalts commented:
This is incredible - removing 18,000 lines of code and tests. We should have an annual award for net lines removed over the year to encourage more of this! Thank you!

bob-carpenter commented:
Before reviewing this, please let me change the absolute delta in finite differences to relative in order to test more extreme values.

syclik commented:
Ok. Please let us know when you want it reviewed.

On Thu, Jun 13, 2019 at 1:11 PM Bob Carpenter <notifications@github.com>
wrote:

> Before reviewing this, please let me change the absolute delta in finite
> differences to relative in order to test more extreme values.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1262?email_source=notifications&email_token=AADH6F5JBXDMUY3FOCLZWODP2J5SLA5CNFSM4HWILOH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXUMORI#issuecomment-501794629>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADH6F355Z4OPGTUWK3NDHTP2J5SLANCNFSM4HWILOHQ>
> .
>

bob-carpenter commented:
This is ready to review now.  I filled out tests to cover existing test cases in rev, fwd, and mix.  We can trim some of these going forward.

rok-cesnovar commented:
This PR will also need a merge with recent develop due to the conflicts due that arised with the merge of the meta flattening.

bob-carpenter commented:
Thanks for the heads up @rok-cesnovar.  I fixed the merge conflict and added some doc.

bob-carpenter commented:
Ack.  Now something broke because of the merge.  This is really annoying when the rug keeps getting pulled out from under my PRs before they're even reviewed.

bob-carpenter commented:
@syclik --- you wanted me to "tag" you on GitHub.  You're already the suggested reviewer, so I didn't know what else to do other than write a message here.

bob-carpenter commented:
@syclik or @seantalts:  I don't understand the Jenkins error.  Everything passed but some upstream failure to build.  Maybe one of you could help me fix the build if it's really an issue.

seantalts commented:
There was an issue then with the Windows EC2 fleet being turned on before it was ready - I'm restarting that phase. /cc @serban-nicusor-toptal

syclik commented:
Thanks, Bob. It’s on my todo to review.

On Wed, Jul 10, 2019 at 2:29 PM seantalts <notifications@github.com> wrote:

> There was an issue then with the Windows EC2 fleet being turned on before
> it was ready - I'm restarting that phase. /cc @serban-nicusor-toptal
> <https://github.com/serban-nicusor-toptal>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1262?email_source=notifications&email_token=AADH6F3Z2WQQFCLBGYW4WTTP6YTB5A5CNFSM4HWILOH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZUK23Q#issuecomment-510176622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADH6FZ3FEEZCWFPSAWDBVLP6YTB5ANCNFSM4HWILOHQ>
> .
>

bob-carpenter commented:
Like I said, it's not critical for 2.20, so I wouldn't
prioritize it over other important things.  I just want
it reviewed sooner rather than later so I don't have
to keep updating and so I can move onto the rest of the
autodiff tests, which are now really easy to write!

- Bob

SteveBronder commented:
@serban-nicusor-toptal it looks like all the other Jenkins instances are waiting on this one? I can't find the button Jenkins has to reset these. Would this happen from us switching from the AWS windows instance from the other windows instance.

https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/activity

Pipeline:
https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/downstream_tests/621/pipeline/

serban-nicusor-toptal commented:
@SteveBronder There was a big queue in Jenkins, it was waiting for its turn.

bob-carpenter commented:
> +  // vv
> +  EXPECT_EQ(f(x1, x2), f(v(x1), v(x2)));
> 
> Well clearly this works, but why do we not need to explicitly extract the val from the var output on the right?
> 
> Like:
> 
> EXPECT_EQ(f(x1, x2), f(v(x1), v(x2)).val());

Because we've implemented operator== for combinations of var and primitive types.

What I'd ask is why EXPECT_EQ rather than EXPECT_FLOAT_EQ---this is testing
floating point with == as is.

rok-cesnovar commented:
@bob-carpenter arent there open PR review comments here?

bob-carpenter commented:
Thanks.  Didn't see that for some reason.  Maybe I get notified for comments and not actual reviews.

> On Aug 5, 2019, at 10:02 AM, Rok Češnovar <notifications@github.com> wrote:
> 
> @bob-carpenter arent there open PR review comments here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented on stan/math/prim/scal/fun/gamma_p.hpp:
```diff
@@ -49,10 +55,23 @@ namespace math {
+   * if x is at a pole of the function
  */
-inline double gamma_p(double x, double a) { return boost::math::gamma_p(x, a); }
+inline double gamma_p(double x, double a) {
```
Done.

bob-carpenter commented on stan/math/prim/scal/fun/gamma_p.hpp:
```diff
@@ -49,10 +55,23 @@ namespace math {
+  if (is_nan(a))
+    return not_a_number();
+  check_positive("gamma_p", "first argument (x)", x);
+  check_positive("gamma_p", "second argument (a)", a);
```
Done.

bob-carpenter commented on test/unit/math/mix/core/operator_divide_equal_test.cpp:
```diff
@@ -1,15 +1,10 @@
-    typename boost::math::tools::promote_args<T1, T2>::type y = x1;
+TEST(mathMixCore, opratorDivideEqual) {
+  auto f = [](const auto& x1, const auto& x2) {
+    decltype(x1 + x2) y = x1;
```
Without the `decltype`, it wouldn't work if `x1` is double and `x2` a `var`, because the `auto y` would infer `double` and then `y /= x2` wouldn't  compile.  There might be an easier way to write this.

bob-carpenter commented on test/unit/math/mix/core/operator_division_test.cpp:
```diff
@@ -1,13 +1,6 @@
-
-TEST(mathMixCore, operatorDivision) {
-  stan::math::test::test_common_args<op_divide_f, false>();
+TEST(mathMixCore, opratorAddition) {
```
Done.

bob-carpenter commented on test/unit/math/mix/core/operator_minus_minus_test.cpp:
```diff
@@ -1,26 +1,36 @@
+  };
+  stan::test::expect_common_unary(f);
+}
+TEST(mathMixCore, opratorMinusMinusPre2) {
```
Because `x1` is const.  So we need a local variable we can apply `--` to and then return the value.  This first functor returns the value before and the second one the value after.  I'll add some doc.

bob-carpenter commented on test/unit/math/mix/core/operator_plus_equal_test.cpp:
```diff
@@ -1,15 +1,10 @@
-  static typename boost::math::tools::promote_args<T1, T2>::type apply(
-      const T1& x1, const T2& x2) {
-    typename boost::math::tools::promote_args<T1, T2>::type y = x1;
+TEST(mathMixCore, opratorPlusEqual) {
```
Found and replaced.

bob-carpenter commented on test/unit/math/mix/core/operator_unary_plus_test.cpp:
```diff
@@ -1,13 +1,6 @@
-
-TEST(mathMixCore, opratorUnaryPlus) {
-  stan::math::test::test_common_args<op_unary_plus_f, false>();
+TEST(mathMixCore, opratorAddition) {
```
Fixed.

bob-carpenter commented on test/unit/math/test_ad.hpp:
```diff
@@ -260,6 +654,216 @@ void expect_ad_vectorized(const F& f, const T1& x1) {
+  typedef fvar<fvar<var>> ffv;
+
+  // vv
+  EXPECT_EQ(f(x1, x2), f(v(x1), v(x2)));
```
Because == works on all of our types.

bob-carpenter commented on test/unit/math/test_ad.hpp:
```diff
@@ -260,6 +654,216 @@ void expect_ad_vectorized(const F& f, const T1& x1) {
+ * Recursive case for variadic unary vectorized function tests for the
+ * specified argument and pack of arguments.
+ *
+ * <p>This test delegates to `expect_ad_vectorized(F, double)`; see that
```
Right.  Fixed doc.

bob-carpenter commented on test/unit/math/test_ad.hpp:
```diff
@@ -1,22 +1,365 @@
+                          double tol = 1e-8) {
+  using stan::math::fabs;
+  if (x1 == 0 && x2 == 0)
+    return;
```
Because if one is zero, the other should be near, but that can't really be absolute tolerance vs. truth as truth is zero.  So there's a separate EXPECT_NEAR case for `(x1 == 0 || x2 == 0)`.  I'll add doc.

bob-carpenter commented on test/unit/math/test_ad.hpp:
```diff
@@ -1,22 +1,365 @@
+ * <p>If both `u` and `v` are zero, the error is defined to be zero.
+ * If only one of `u` or `v` is zero, the above definition always
+ * yields 1.  This is not a useful test, so we instead test absolute
+ * error in the case that one of the values is zero.
```
OK.

bob-carpenter commented on test/unit/math/mix/mat/fun/digamma_test.cpp:
```diff
@@ -1,98 +1,7 @@
+TEST(mathMixMatFun, digamma) {
+  auto f = [](const auto& x1) { return stan::math::digamma(x1); };
+  stan::test::expect_common_nonzero_unary_vectorized(f);
+  stan::test::expect_unary_vectorized(f, -25, -10.2, -1.2, -1, 2.3, 5.7);
```
Yes.  It'll just instantiate a generic type `T` that just keeps getting passed along down to `expect_ad_helper`, where it gets passed directly to the functor `f`. I can add a test that makes it clear that this arduous path is actually working.

bbbales2 replied:
Yeah would it be possible to add a few tests to make sure the right template deduction is happening in these automatic types and things aren't getting cast away?

Like make sure 1 is an int, 1.0 is a double, var stays var? Is this what you mean by the arduous path?

bob-carpenter replied:
Types get preserved because it's just templated arguments all the way down the call chain until the function being tested is invoked, and that's defined with a template parameter in order to match primitive and autodiff types.

I'll add some tests of the test framework to demonstrate from the user-callable test functions (I should split this into an internal helper file and what a developer's going to troll through to find the right test function).

The buck stops here, though---I don't intend to test the tests of the test framework. :-)

bbbales2 replied:
> I can add a mock function that throws on integer values and check directly that it throws on int and not on double. Will that be enough?

Yeah. If I was writing tests, I would probably write some code to convince myself the correct thing was happening. Just too easy to promote to double.

If I know that this behavior is being checked for me I can rest easy.

bob-carpenter replied:
All tested now.  And the tests are tested (though you have to edit them to add in a test that fails to show that failure happens).  I didn't want to start thinking about how to start testing the Google test framework macros.

bob-carpenter replied:
Now onto part two.  

> So we expect all the error checking parts of functions to be checked in the prim checks? So that we can trust the prim implementations of the functions?

Yes.  These tests all assume the primitive versions return correct values and throw in the right places.  This framework does *not* test that.

> In that case, these functions were missing prim tests. Not sure I got all of them: acos, asin, atan, ceil, cos, fabs, floor, log10, sin, sinh, tanh

I'll add a bunch of automatic tests that make sure that `stan::math::F` returns same values and throws in same places as `std::F` for all the functions `F` we implement to the extent possible.  There's going to be trouble with `stan::math::log1p` because the behavior's been changed to not conform to `std::log1p`.   I might find other cases of that, in which case I obviously can't test vs `std::`.  I don't want this PR to degrade into testing or whole math lib!




ceil and floor, at least, should have tests. Want to add those in this pull or make another issue? I'd personally prefer to get them implemented (or found in case I missed them) now.

I need to test a function or two myself to understand the limits of this stuff. Do you see any difficulty testing more complicated stuff?

The matrix stuff, for instance? Like a cholesky, or matrix multiply, or something?

bob-carpenter commented on test/unit/math/mix/mat/fun/trigamma_test.cpp:
```diff
@@ -1,102 +1,7 @@
+TEST(mathMixMatFun, trigamma) {
+  auto f = [](const auto& x1) { return stan::math::trigamma(x1); };
+  // stan::test::expect_common_unary_vectorized(f);
+  stan::test::expect_unary_vectorized(f, -0.5, 0.5);
```
The previous tests didn't test nearly as deeply.  When I put those back, finite diffs are off by just a hair more than tolerance.  

I'm going to have to add a way to pass a factor through this whole pipeline to lower tolerances.  I was thinking I could get away with one set for everything, but obviously not.

bob-carpenter replied:
Whew 2.  I fished tolerances all the way through the test framework so I could replace the missing trigammas.  I had to loosen tolerances a bit, but at least this demonstrates that new part of the framework works.

bob-carpenter commented:
I'm working on fixing this by making tolerances configurable from the top-level calls.  I'm also adding doc for some of the confusing points in the code that @bbbales2 was asking about in the review.  I'll write a Wiki page on how to use all this from the top-level, too.  I really want it to be general enough that we can do all of our testing through this interface.

bob-carpenter commented:
How do I test primitive functions like `cos`?  I think so far we've just been assuming the standard library gets them right.  Would it be sufficient to add tests that they get results near the standard library results so that we'll catch bad regressions?

bob-carpenter commented:
> I need to test a function or two myself to understand the limits of this stuff. Do you see any difficulty testing more complicated stuff?

See: `test/unit/math/test_ad_test.cpp`.

> The matrix stuff, for instance? Like a cholesky, or matrix multiply, or something?

There's a test of matrix time scalary mutliply just to show it could deal with heterogeneous types and multivariate outputs.

> The matrix stuff, for instance? Like a cholesky, or matrix multiply, or something?

No, there's a test for matrix times scalar in `  Everything else will work the same way.

bbbales2 commented:
For ceil and floor, just a couple tests each to make sure they behave like our doc says they should. Don't some languages do different things for negative numbers? That's the reason I think they should just have rudimentary tests.

For the rest I think just painfully simple tests are good. Like just one EXPECT_EQ(stan::math::cos(5.0), std::cos(5.0)) is more than sufficient. I'm happy to merge without these too -- only calling them out since the rest of the library is so carefully checked.

I don't think we need to do inf/nan testing and all the complex stuff.

bbbales2 commented:
Edit: If there isn't any ambiguity on ceil/floor implementations in the wild, I'm happy to leave those out too.

bob-carpenter commented:
I'll write more general tests for the library functions.  I have all the basic comparisons to deal with inf and NaN now.

Turns out you were right---those `int` arguments get mangled into `double` way down in the bowels of serialize and deserialize, which uses an Eigen vector.  Trying to write something that threw on `int` showed it never got called.  

I'm now a little unclear on what to test.  Integer inputs should just behave like floating point inputs wherever there is autodiff possible.  Would it be enough to test values in that case?  The problem I see is all the integer and double interaction.  The current tests are very thin on testing integer behavior, but that's no excuse not to try to figure out how to do it right.

Anyway, back to the drawing board on this one, I'm afraid.

This is going to be a huge pain if we need to test mixed arguments of int and real.

bbbales2 commented:
I don't know if this could be co-opted into serialize/deserialize, but this is the thing I used on the last pull: https://github.com/stan-dev/math/blob/4ba819b5b27f9a0616fe86a4618a581ac00b6160/stan/math/prim/mat/fun/variable_adapter.hpp

Have a look at the first couple tests here: https://github.com/stan-dev/math/blob/4ba819b5b27f9a0616fe86a4618a581ac00b6160/test/unit/math/prim/mat/fun/variable_adapter_test.cpp

```make_argument_adapter<T>(args...)``` pulls out all the arguments in args of scalar type T and exposes them in a 1d interface. You can do whatever you want with them (using ```operator()``` and ```size()``` member functions), and then call ```get_args()``` to get things back as a tuple (which can then be used to call the function with apply).

Not sure if that works here or not.

I'm also not sure about the exact mixing of ints and doubles and such we want tested. I feel like I remember this causing a couple bugs before though I'm not sure.

bob-carpenter commented:
Yes, this ahs caused bugs before, primarily due to integer division. For example, if you send in the three arguments to `normal_lpdf()` as `int`, you don't want it to execute `(y - mu) / sigma` (it doesn't---it puts everything in views).

I'll take a look at what you had before.  It may be better than the hack I came up with last night, which is to use traits to test if something is assignable to `int` and if the round trip doesn't change value, then testing with both `int` and `double` args.  Everywhere `real` is allowed in Stan should also allow `int`, so this is a good test to have in place.

bob-carpenter commented:
Thanks by the way, for such a thorough review.  I'm especially grateful given that it's replacing something you spent a ton of work on.

bob-carpenter commented:
Whew.  I got the testing sorted for integers.  The trick was to specialize the `expect_ad_X` functions to integers.  Obviously you can't autodiff an `int`, so I'm just testing (1) the value is the same, and (2) autodiff works for the `int` cast to `double, and (3) for binary functions with one `int` arg, also binding the `int` to test the resulting unary function.  The specialization code is relatively small because it delegates the hard tests recursively and can use a closure to bind the `int`. 

I had to generalize the basic expect tests, but I think they're good enough now that we could export them for general use if we wanted to in the testing framework.  The only rough part relative to Google test is error message construction.  Maybe they'd need to be generalized to allow the really convenient `TEST(...) << a1 << ... << aN;` pattern.  I'm not quite sure how to write a good enough macro to do that.

bob-carpenter commented on test/unit/math/mix/mat/fun/Phi_test.cpp:
```diff
@@ -1,103 +1,9 @@
+  stan::test::expect_common_unary_vectorized(f);
+  stan::test::expect_unary_vectorized(f, -27.5, 27.5);
+  for (double x = -37.5; x <= 10; x += 0.5)
+    stan::test::expect_unary_vectorized(x);
```
Testing all the values between -37.5 and 10.  That -27.5 above is redundant.  The vectorized tests test that the function works and all the vectorized forms.

bob-carpenter commented:
@bbbales2:  All wrapped up and ready to review.  I addressed all of the comments and marked the conversations closed where possible.  The big changes since the original PR are all [reflected in the revised autodiff Wiki](https://github.com/stan-dev/math/wiki/Automatic-Differentiation-Testing-Framework).  Here's the list.

* Integers are now passed properly and I added integers to the common tests that most functions get.  When an `int` argument is supplied to a test, it tests that the value and exceptions are the same as for `double`, and that autodiff works at the `int` cast to `double`.  There are now tests that can be edited out in `test_ad_test.cpp` that will fail where they should when `int` behavior doesn't match `double` behavior.

* Tolerances may be provided from all top-level calls (this allowed me to add back in missing tests for `trigamma`).  There is now a struct with a default tolerance that can be reset for every single value returned by our functionals for gradients, Hessians, and gradients of Hessians.  

* There's now a framework to test that two primitive functions match.  I added a bunch of tests for Stan functions that have `std::` equivalent to make sure value and exception behavior matches.  
    > *Warning:*  This can't be done everywhere because functions like `stan::math::log1p` now no longer match `std::log1p` in exception behavior--the Stan version throws where `log1p` returns NaN.  I didn't feel it within the scope of this PR to change that behavior.

I see some remaining issues for testing that I didn't try to tackle in this PR:

1.  Poles.  We have functions in the math lib that define derivatives based on one-sided limits or just ad hoc decisions (like for `fabs(0)`) that can't be tested with finite diffs.  We could add one-sided finite diff functionals (does anyone know formulas for that?).

2.  Matching standard library behavior.  We now have functions like `stan::math::log1p` which throw domain exceptions where `std::log1p` does not.

3.  Probabiltiy functions.  These still use their own testing framework.  It should be consolidated with this one to test the reducing vectorization we do in those functions.

4.  Other functions.  I removed two existing test frameworks and consolidated them into this one (the specialized one for vectorized functions and the generic scalar one I'd created for basic functions).  But I only added the new thorough tests for functions that were being tested by one of those frameworks previously. 

When (4) is done (others don't need to be done), we'll be ready to officially "release" higher-order autodiff.  Then we'll want:

* An *arXiv* paper like our first one explaining how it works.  It'd be nice to write a combined paper that's more concise, too.    A lot has changed since our first autodiff system with additional solvers (algebraic and ode), GPU and MPI and the way we manage memory.  Measured in least publishable units (LPUs), there's a gold mine here!

bob-carpenter commented on test/unit/math/ad_tolerances.hpp:
```diff
@@ -0,0 +1,66 @@
+ * 1e-3 for second derivatives, and 1e-2 for third derivatives.  The
+ * names begin with the functional being evaluated, include an `fvar`
+ * if the function is implemented using only forward mode, and end
+ * with the quantity being calculated; for exmaple,
```
Oops.  Missed that one.  I have it updated locally but not pushed.  I didn't want to kick off another round of tests fora  typo.  Our whole doc repo could use spell checking.  I just banged a lot of it out typing as fast as I could.

bob-carpenter replied:
fixed.

bob-carpenter commented on test/unit/math/serializer_test.cpp:
```diff
@@ -0,0 +1,66 @@
+  EXPECT_EQ(-10, y(0));
+  EXPECT_EQ(-20, y(1));
+  Eigen::RowVectorXd z = d.read(Eigen::RowVectorXd(3));
+  EXPECT_EQ(-10, y(0));
```
Fixed.  Looks like I will have to push.  I missed a bunch of these little ones before declaring victory.

bob-carpenter commented:
> can this [exception contents] be tested using this: google/googletest#952 (comment) ?

The problem is capturing type of exception generically.  We're trying to keep our math lib down to just domain exceptions and indexing errors, I think.  I've leaned on some of this in `rethrow_located` in the language.

So I don't know how to do better than what we're doing now.  Testing messages would be possible, but there's no reason to expect the primitive and `var` version of a function to throw exactly the same text in an exception.

bob-carpenter commented:
> For ceil and floor, just a couple tests each to make sure they behave like our doc says they should. Don't some languages do different things for negative numbers? That's the reason I think they should just have rudimentary tests.

> For the rest I think just painfully simple tests are good.

I added pretty thorough tests to match values and exceptions for all of them (other than `log1p`, which went its own way with exception behavior).

andrjohns commented:
> I added pretty thorough tests to match values and exceptions for all of them (other than `log1p`, which went its own way with exception behavior).

For functions like ```log1p``` that handle exceptions differently to other ```std```, this was a leftover from the change from ```boost::math``` to ```std```. Since the boost functions throw on exceptions, the agreement was that the ```std``` implementations should as well.

[The decision/discussion was in this thread](https://github.com/stan-dev/math/pull/1067#issuecomment-443400498)
and
[The list of functions where this was done is in this post](https://github.com/stan-dev/math/issues/1010#issuecomment-445527864)

bob-carpenter commented:
@bbbales2 : If this passes, it's ready for final review.  Doing these thorough integer tests did uncover one bug in the binary test framework for int, int inputs, and that's fixed now.

I left the commented-out tests in because I think of the unit tests as also a kind of developer doc.  They show what failing tests look like and those can be run if you edit them out.  If you think I should get rid of them, I will.

bbbales2 commented:
All good!
